skbio/diversity/beta/_unifrac.py

```diff
@@ -0,0 +1,170 @@
+        raise ValueError("Input vectors u_counts, v_counts, and otu_ids must"
+                         " be of equal length.")
+    u_observed_otus = {o: c for o, c in zip(otu_ids, u_counts) if c >= 1}
+    v_observed_otus = {o: c for o, c in zip(otu_ids, v_counts) if c >= 1}
+    observed_nodes1 = set(tree.observed_node_counts(u_observed_otus))
+    observed_nodes2 = set(tree.observed_node_counts(v_observed_otus))
+    observed_branch_length = sum(
+        o.length for o in observed_nodes1 | observed_nodes2
+        if o.length is not None)
+    if observed_branch_length == 0:
+        # boundary case where both communities have no members
+        return 0.0
+    shared_branch_length = sum(
+        o.length for o in observed_nodes1 & observed_nodes2
+        if o.length is not None)
+    unique_branch_length = observed_branch_length - shared_branch_length
+    unweighted_unifrac = unique_branch_length / observed_branch_length
+    return unweighted_unifrac
+
+
+@experimental(as_of="0.4.0-dev")
+def weighted_unifrac(u_counts, v_counts, otu_ids, tree, normalized=False):
+    """ Compute weighted unifrac with or without branch length normalization
+
+    Parameters
+    ----------
+    u_counts, v_counts: list, np.array
+        Vectors of counts of OTUs for two samples. Must be equal length.
+    otu_ids: list, np.array
+        Vector of OTU ids corresponding to tip names in ``tree``. Must be the
+        same length as ``u_counts`` and ``v_counts``.
+    tree: skbio.TreeNode
+        Tree relating the OTUs in otu_ids. The set of tip names in the tree can
+        be a superset of ``otu_ids``, but not a subset.
+    normalized: boolean, optional
+        If ``True``, apply branch length normalization. Resulting distances
+        will then be in the range ``[0, 1]``.
+
+    Returns
+    -------
+    float
+        The weighted unifrac distance between the two samples.
+
+    Raises
+    ------
+    ValueError
+        If ``u_counts``, ``v_counts``, and ``otu_ids`` are not all equal in
+        length.
+    MissingNodeError
+        If an OTU id is provided that does not correspond to a tip in the
+        tree.
+
+    See Also
+    --------
+    unweighted_unifrac
+
+    Notes
+    -----
+    Unweighted unifrac was originally described in [1]_, which includes a
+    discussion of unweighted (qualitative) versus weighted (quantitiative)
+    diversity metrics.
+
+    References
+    ----------
+    .. [1] Lozupone, C. A., Hamady, M., Kelley, S. T. & Knight, R. Quantitative
+       and qualitative beta diversity measures lead to different insights into
+       factors that structure microbial communities. Appl. Environ. Microbiol.
+       73, 1576–1585 (2007).
+
+    """
+    u_observed_otus = {o: c for o, c in zip(otu_ids, u_counts) if c >= 1}
+    u_total_count = sum(u_counts)
+    v_observed_otus = {o: c for o, c in zip(otu_ids, v_counts) if c >= 1}
+    v_total_count = sum(v_counts)
+    if u_total_count == 0 and v_total_count == 0:
```

Comment:
since a negative count doesn't make sense, doesn't that imply that this test could be done first and with `if not len(u_counts) and not len(v_counts)`?